oracle and randomness overhaul (#551)
loomnetwork/cryptozombie-lessons: en/20/02.md

---
title: Importing from NPM and Github
actions: ['checkAnswer', 'hints']
material:
  editor:
    language: sol
    startingCode:
      "PriceConsumerV3.sol": |
        pragma solidity ^0.6.7;

        contract PriceConsumerV3 {

        }
    answer: |
      pragma solidity ^0.6.7;

      import "@chainlink/contracts/src/v0.6/interfaces/AggregatorV3Interface.sol";

      contract PriceConsumerV3 {

      }
---

> Now to have our contract pull the price from another contract, we need it's interface / ABI. Although... where we currently are, we aren't sure what the interface of a Chainlink Data Feed contract is? How do we get that? 

Reviewer: ```suggestion
Now to have our contract pull the price from another contract, we need its interface / ABI. Although... where we currently are, we aren't sure what the interface of a Chainlink Data Feed contract is? How do we get that? 
```

## Importing from NPM / Github

Did you know, you can actually import code from outside your contracts? I know it's amazing right! Often times you don't have to have every piece of code directly in your project, you can borrow it from other applications!

> We want to import the `AggregatorV3Interface` from the [Chainlink github repository](https://github.com/smartcontractkit/chainlink). This interface will include all the functions that we need to interace with a Chainlink Data Feed contract. This will include functions like `latestRoundData()` which will return all the information we need. 

Reviewer: ```suggestion
We want to import the `AggregatorV3Interface` from the <a href="https://github.com/smartcontractkit/chainlink" target=_new>Chainlink GitHub repository</a>. This interface will include all the functions that we need to interact with a Chainlink Data Feed contract, including functions like `latestRoundData()` which will return all the information we need. 
```

Here is what the full interface looks like: 
```javascript
// SPDX-License-Identifier: MIT
pragma solidity ^0.6.0;

interface AggregatorV3Interface {

  function decimals()
    external
    view
    returns (
      uint8
    );

  function description()
    external
    view
    returns (
      string memory
    );

  function version()
    external
    view
    returns (
      uint256
    );

  // getRoundData and latestRoundData should both raise "No data present"
  // if they do not have data to report, instead of returning unset values
  // which could be misinterpreted as actual reported values.
  function getRoundData(
    uint80 _roundId
  )
    external
    view
    returns (
      uint80 roundId,
      int256 answer,
      uint256 startedAt,
      uint256 updatedAt,
      uint80 answeredInRound
    );

  function latestRoundData()
    external
    view
    returns (
      uint80 roundId,
      int256 answer,
      uint256 startedAt,
      uint256 updatedAt,
      uint80 answeredInRound
    );

}
```

> And we can find it in the [github repo for AggregatorV3Interface.](https://github.com/smartcontractkit/chainlink/blob/master/contracts/src/v0.6/interfaces/AggregatorV3Interface.sol)

Reviewer: ```suggestion
And we can find it in the <a href="https://github.com/smartcontractkit/chainlink/blob/master/contracts/src/v0.6/interfaces/AggregatorV3Interface.sol" target=_new>GitHub repo for AggregatorV3Interface</a>.
```

> We can either import directly from Github, or from [NPM packages](https://www.npmjs.com/package/@chainlink/contracts) to get this contract. The framework you're using (like Truffle, Brownie, Remix, Hardhat) will determine whether or not to use github or NPM packages, but the syntax will be approximately the same!

Reviewer: ```suggestion
We can either import directly from GitHub, or from <a href="https://www.npmjs.com/package/@chainlink/contracts" target=_new>NPM packages</a> to get this contract. The framework you're using (like Truffle, Brownie, Remix, Hardhat) will determine whether or not to use GitHub or NPM packages, but the syntax will be approximately the same!
```
